flocker/control/httpapi.py

```diff
@@ -466,10 +469,31 @@ def create_container_configuration(self, host, name, image):
         # Find the node.
         node = self._find_node_by_host(host, deployment)
 
+        # Check if we have any ports in the request. If we do, check existing
+        # external ports exposed to ensure there is no conflict. If there is a
+        # conflict, return an error.
+
+        if ports:
+            for port in ports:
+                for application in node.applications:
+                    for application_port in application.ports:
+                        if application_port.external_port == port['external']:
+                            raise CONTAINER_PORT_COLLISION
+
         # Create Application object, add to Deployment, save.
         application = Application(name=name,
                                   image=DockerImage.from_string(image))
 
+        # If we have ports specified, add these to the Application instance.
+        if ports:
+            application_ports = []
+            for port in ports:
+                application_ports.append(Port(
+                    internal_port=port['internal'],
+                    external_port=port['external']
+                ))
+            application.ports = frozenset(application_ports)
```

Comment:
I am going to make `Application` uses pyrsistent, so it's going to be immutable. I suggest building up a dictionary of additional keyword arguments to pass to `Application()`.